sis/sis_lib/help/sis.1
.TH SIS 1CAD "1 July 1994"
.SH NAME
sis \- Sequential Interactive System
.SH SYNOPSIS
.B sis
[options] [file]
.SH DESCRIPTION
\fBSIS\fR is an algorithmic sequential circuit optimization
program.  \fBSIS\fR starts from a description
of a sequential logic macro-cell and produces
an optimized set of logic equations plus latches which preserves
the input-output behavior of the macro-cell.
The sequential circuit can be stored as a finite
state machine or as an implementation consisting of logic
gates and memory elements.
The program
includes algorithms for minimizing the area required to
implement the logic equations, algorithms for minimizing delay,
and a technology
mapping step to map a network into a user-specified cell library.
It includes all of the optimization techniques available in
\fBMIS\fR, and replaces \fBMIS\fR completely.
.PP
\fBSIS\fR can be run in interactive mode accepting commands
from the user, or in batch mode reading commands
from a file or from the command line.
If no options are given on the command line, \fBSIS\fP will
enter interactive mode.
Otherwise, \fBSIS\fP will enter batch mode.
When running in batch mode, \fBSIS\fP reads its
input from the file given on the command line, or from standard input
if no filename is given;
output is directed to standard output, unless \fB-o\fP is used to
specify an output filename.
.PP
When \fBSIS\fP starts-up, it performs an initial source of the
files $SIS/sis_lib/.misrc and $SIS/sis_lib/.sisrc.
Typically this defines a standard
set of aliases for various commands.
Following that the files ~/.misrc, ~/.sisrc, ./misrc, and ./sisrc
are sourced for user-defined aliases at startup.
.SH OPTIONS
.TP 8
.B -c cmdline
Run \fBSIS\fR in batch mode, and execute \fBcmdline\fR.
Multiple commands are separated with semicolons.
.TP 8
.B -f script
Run \fBSIS\fR in the batch mode, and execute commands from the
file \fBscript\fR.
.TP 8
.B -t type
Specifies the type of the input when running in batch mode.
The legal input types are:
Berkeley Logic Interchange Format (\fB-t blif\fP),
eqntott(1CAD)-format equation input (\fB-t eqn\fP),
\fIKISS2 format\fR (\fB-t kiss\fP),
\fIOct Logic View\fR (\fB-t oct\fP),
Berkeley PLA Format (\fB-t pla\fP),
\fISLIF format\fR (\fB-t slif\fP),
and suppress input (\fB-t none\fP).
The default input type is \fIblif\fR.
.TP 8
.B -T type
Specifies the type of the output when running in batch mode.
The legal output types are:
bdnet(1CAD)-format net-list (\fB-T bdnet\fP),
Berkeley Logic Interchange Format (\fB-T blif\fP),
eqntott(1CAD)-format equation input (\fB-T eqn\fP),
\fIKISS2 format\fR (\fB-T kiss\fP),
\fIOct logic view\fR (\fB-T oct\fP),
Berkeley PLA Format (\fB-T pla\fP),
\fISLIF format\fR (\fB-T slif\fP),
and suppress output (\fB-T none\fP).
The default output type is \fIblif\fR.
.TP 8
.B -o file
Specifies the output file when running in batch mode.
For \fIOct\fP output, this is a string of the
form \fIcell\fP:\fIview\fP.
The default for the output is the standard output.
.TP 8
.B -s
Suppress sourcing the commands from the standard
startup script ($SIS/sis_lib/.misrc and $SIS/sis_lib/.sisrc).
.TP 8
.B -x
For batch mode operation, suppress reading an initial network, and
suppress writing an output network.
Equivalent to \fB-t none -T none\fP.
.de XX
.sp 1
.LP
.ti -5
.B
..
.de X1
.br
.ti -5
.B
..
.PP
.SH COMMAND SUMMARY
All commands are summarized below according
to their function : network manipulation (operations
on the logic-level implementation),
ASTG manipulation (operations on the asynchronous
signal transition graph), STG manipulation (operations
on the synchronous state transition graph),
input-output, network status, command interpreter,
and miscellaneous.
The last two tables summarize the newest commands that
operate on ASTG's and sequential circuits, respectively.
.sp 2
.TS
.if \n+(b.=1 .nr d. \n(.c-\n(c.-1
.de 35
.ps \n(.s
.vs \n(.vu
.in \n(.iu
.if \n(.u .fi
.if \n(.j .ad
.if \n(.j=0 .na
..
.nf
.nr #~ 0
.if n .nr #~ 0.6n
.ds #d .d
.if \(ts\n(.z\(ts\(ts .ds #d nl
.fc
.nr 33 \n(.s
.rm 80 81
.nr 80 0
.nr 38 \wact_map
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \wadd_inverter
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \wastg_slow
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \wastg_syn
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \wastg_to_f
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \wastg_to_stg
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \wbuffer_opt
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \wchng_clock
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \wcollapse
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \wdecomp
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \weliminate
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \wenv_seq_dc
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \wequiv_nets
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \wespresso
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \wextract_seq_dc
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \wfactor
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \wfanout_alg
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \wfanout_param
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \wfree_dc
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \wforce_init_0
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \wfull_simplify
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \wfx
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \wgcx
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \wgkx
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \winvert
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \wlatch_output
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \wmap
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \wone_hot
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \wphase
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \wred_removal
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \wreduce_depth
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \wremove_dep
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \wremove_latches
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \wreplace
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \wresub
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \wretime
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \wsimplify
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \wspeed_up
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \wspeed_up_alg
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \wstate_assign
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \wstg_extract
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \wstg_to_network
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \wsweep
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \wtech_decomp
.if \n(80<\n(38 .nr 80 \n(38
.nr 38 \wwd
.if \n(80<\n(38 .nr 80 \n(38
.80
.rm 80
.nr 81 0
.nr 38 \wtechnology mapping to Actel architecture
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \wadd inverters to the network to make all gates negative
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \wremove hazards from the ASTG implementation
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \w(uses bounded wire delay model)
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \wsynthesize a two-level implementation from the ASTG
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \w(uses unbounded wire delay model)
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \wgenerate a two-level implementation of each output of the ASTG
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \w(uses bounded wire delay model)
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \wgenerate an STG from the ASTG
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \winserts buffering trees for high fanout gates
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \wtoggles clock setting between user-specification and generated values
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \wcollapse a network or a set of nodes
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \wdecompose a node into a set of nodes
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \weliminates nodes whose value falls below a threshold
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \wextract sequential don't cares based on the environment
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \wgroup and merge nets by equivalence classes
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \wcollapse network and minimize with a two-level minimizer
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \wextract sequential don't cares
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \wdetermine a factored form for a node
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \wselect a fanout optimization algorithm (to be used by map)
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \wset some parameters for fanout algorithm (to be used by map)
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \wfrees the external don't care network
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \wmodify so all latches to have a 0 initial state
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \wsimplify the nodes using local compatible don't cares
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \wdo fast extraction of the best double cube and single cube divisors
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \wextract common cubes from the network
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \wextract common multiple-cube divisors from the network
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \winvert a node, and toggle the phase of all of its fanouts
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \wforces some outputs to be fed directly by latches
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \wtechnology mapping to find an implementation for the network
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \wquick one-hot encoding
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \wphase assignment to minimize number of inverters
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \wperform redundancy removal via atpg
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \wincrease the speed before mapping by reducing the depth
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \wremoves some structural (but not logical) dependencies
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \wremoves redundant latches
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \wquick algebraic decomposition on 2-input NANDs
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \wperform resubstitution of a node into other nodes in the network
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \wmove the latches in the circuit to minimize cycle time/# latches
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \wtwo-level minimization of each node
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \wrestructure critical paths to reduce delay
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \wseveral algorithms for performance enhancement
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \wcreate the logic from the STG using state assignment
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \wextract an STG from the logic
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \wconverts a state-encoded STG to a logic network
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \wremove all inverters, buffers, and unnecessary latches from the network
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \wdecompose a network for technology mapping
.if \n(81<\n(38 .nr 81 \n(38
.nr 38 \wre-express a node using another node using weak division
.if \n(81<\n(38 .nr 81 \n(38
.81
.rm 81
.nr 38 \w\fBNetwork Manipulation Commands\fP-\n(80-3n-\n(81
.if \n(38>0 .nr 38 \n(38/1
.if \n(38<0 .nr 38 0
.nr 81 +\n(38
.nr 38 1n
.nr 79 0
.nr 40 \n(79+(1*\n(38)
.nr 80 +\n(40
.nr 41 \n(80+(3*\n(38)
.nr 81 +\n(41
.nr TW \n(81
.nr TW +1*\n(38
.if t .if \n(TW>\n(.li .tm Table at line 178 file Input is too wide - \n(TW units
.ne 49v+2p
.nr #I \n(.i
.in +(\n(.lu-\n(TWu-\n(.iu)/2u
.fc  
.nr #T 0-1
.nr #a 0-1
.nr #a 0-1
.eo
.de T#
.ds #d .d
.if \(ts\n(.z\(ts\(ts .ds #d nl
.mk ##
.nr ## -1v
.ls 1
.if \n(#T>=0 .nr #a \n(#T
.if \n(T. .vs \n(.vu-\n(.sp
.if \n(T. \h'|0'\s\n(33\l'|\n(TWu\(ul'\s0
.if \n(T. .vs
.if \n(#a>=0 .sp -1
.if \n(#a>=0 \h'|0'\s\n(33\h'-\n(#~u'\L'|\n(#au-1v'\s0\v'\n(\*(#du-\n(#au+1v'\h'|\n(TWu'
.if \n(#a>=0 .sp -1
.if \n(#a>=0 \h'|\n(TWu'\s\n(33\h'-\n(#~u'\L'|\n(#au-1v'\s0\v'\n(\*(#du-\n(#au+1v'
.ls
..
.ec
.nr 36 \n(.v
.vs \n(.vu-\n(.sp
\h'|0'\s\n(33\l'|\n(TWu\(ul'\s0
.vs \n(36u
.mk #a
.ta \n(81u 
.nr 31 \n(.f
.nr 35 1m
\&\h'|\n(40u'\fBNetwork Manipulation Commands\fP
.nr 36 \n(.v
.vs \n(.vu-\n(.sp
\h'|0'\s\n(33\l'|\n(TWu\(ul'\s0
.vs \n(36u
.ta \n(80u \n(81u 
.nr 31 \n(.f
.nr 35 1m
\&\h'|\n(40u'act_map\h'|\n(41u'technology mapping to Actel architecture
.ta \n(80u \n(81u 
.nr 31 \n(.f
.nr 35 1m
\&\h'|\n(40u'add_inverter\h'|\n(41u'add inverters to the network to make all gates negative
.ta \n(80u \n(81u 
.nr 31 \n(.f
.nr 35 1m
\&\h'|\n(40u'astg_slow\h'|\n(41u'remove hazards from the ASTG implementation
.ta \n(80u \n(81u 
.nr 31 \n(.f
.nr 35 1m
\&\h'|\n(40u'\h'|\n(41u'(uses bounded wire delay model)
.ta \n(80u \n(81u 
.nr 31 \n(.f
.nr 35 1m
\&\h'|\n(40u'astg_syn\h'|\n(41u'synthesize a two-level implementation from the ASTG
.ta \n(80u \n(81u 
.nr 31 \n(.f
.nr 35 1m
\&\h'|\n(40u'\h'|\n(41u'(uses unbounded wire delay model)
.ta \n(80u \n(81u 
.nr 31 \n(.f
.nr 35 1m
\&\h'|\n(40u'astg_to_f\h'|\n(41u'generate a two-level implementation of each output of the ASTG
.ta \n(80u \n(81u 
.nr 31 \n(.f
.nr 35 1m
\&\h'|\n(40u'\h'|\n(41u'(uses bounded wire delay model)
.ta \n(80u \n(81u 
.nr 31 \n(.f
.nr 35 1m
\&\h'|\n(40u'astg_to_stg\h'|\n(41u'generate an STG from the ASTG
.ta \n(80u \n(81u 
.nr 31 \n(.f
.nr 35 1m
\&\h'|\n(40u'buffer_opt\h'|\n(41u'inserts buffering trees for high fanout gates
.ta \n(80u \n(81u 
.nr 31 \n(.f
.nr 35 1m
\&\h'|\n(40u'chng_clock\h'|\n(41u'toggles clock setting between user-specification and generated values
.ta \n(80u \n(81u 
.nr 31 \n(.f
.nr 35 1m
\&\h'|\n(40u'collapse\h'|\n(41u'collapse a network or a set of nodes
.ta \n(80u \n(81u 
.nr 31 \n(.f
.nr 35 1m
\&\h'|\n(40u'decomp\h'|\n(41u'decompose a node into a set of nodes
.ta \n(80u \n(81u 
.nr 31 \n(.f
.nr 35 1m
\&\h'|\n(40u'eliminate\h'|\n(41u'eliminates nodes whose value falls below a threshold
.ta \n(80u \n(81u 
.nr 31 \n(.f
.nr 35 1m
\&\h'|\n(40u'env_seq_dc\h'|\n(41u'extract sequential don't cares based on the environment
.ta \n(80u \n(81u 
.nr 31 \n(.f
.nr 35 1m
\&\h'|\n(40u'equiv_nets\h'|\n(41u'group and merge nets by equivalence classes
.ta \n(80u \n(81u 
.nr 31 \n(.f
.nr 35 1m
\&\h'|\n(40u'espresso\h'|\n(41u'collapse network and minimize with a two-level minimizer
.ta \n(80u \n(81u 
.nr 31 \n(.f
.nr 35 1m
\&\h'|\n(40u'extract_seq_dc\h'|\n(41u'extract sequential don't cares
.ta \n(80u \n(81u 
.nr 31 \n(.f
.nr 35 1m
\&\h'|\n(40u'factor\h'|\n(41u'determine a factored form for a node
.ta \n(80u \n(81u 
.nr 31 \n(.f
.nr 35 1m
\&\h'|\n(40u'fanout_alg\h'|\n(41u'select a fanout optimization algorithm (to be used by map)

ChangeLog
ChangeLog
---------

2005-03-18 Release 1.3.6
2005-03-08 Added proper but still incomplete AUTHORS list
2005-03-07 Added support for CUDD 2.4.0
2005-03-07 Fixed configure to do subdirs correctly
2005-03-07 Some compile warnings cleaned up

2004-05-05 Release 1.3.5
2004-05-05 Added sis_lib/.misrc (Ken McMillan/Ellen Sentovich)
2004-05-05 Fixed bug in eliminate (Ken McMillan/Ellen Sentovich)

2004-03-27 Release 1.3.4
2004-03-27 Fixed for Cygwin

2004-03-16 Release 1.3.3
2004-03-16 Fixed fclose of null FILE * in power
2004-03-11 Now uses autoconf/automake

2004-02-18 Release 1.3.2
2004-02-18 Mac OSX compilation fixes

2004-02-11 Release 1.3.1
2004-02-11 Copyright notice added
2004-02-09 Fixes to ease compilation

2004-02-08 Release 1.3


sis/main/sis.c
#include "sis.h"

#include <sys/stat.h>

#include <readline/readline.h>
#include <readline/history.h>

extern void init_sis();
extern void end_sis();

static int main_has_restarted = 0;
static network_t *network; /* allows for restart ... */

char **command_name_completion(const char *, int, int);
char *command_name_generator(const char *, int);
char *commands[] = {
	"act_map",
	"add_inverter",
	"alias",
	"astg_add_state",
	"astg_contract",
	"astg_current",
	"astg_encode",
	"astg_lockgraph",
	"astg_marking",
	"astg_persist",
	"astg_print_sg",
	"astg_print_statastg_slow",
	"astg_state_min",
	"astg_stg_scr",
	"astg_syn",
	"astg_to_f",
	"astg_to_stg",
	"atpg",
	"bdsyn",
	"buffer_opt",
	"c_check",
	"c_opt",
	"chng_clock",
	"chng_name",
	"collapse",
	"constraints",
	"decomp",
	"echo",
	"eliminate",
	"env_seq_dc",
	"env_verify_fsm",
	"equiv_nets",
	"espresso",
	"extract_seq_dc",
	"factor",
	"fanout_alg",
	"fanout_param",
	"force_init_0",
	"free_dc",
	"full_simplify",
	"fx",
	"gcx",
	"gkx",
	"help",
	"history",
	"invert",
	"invert_io",
	"ite_map",
	"latch_output",
	"map",
	"one_hot",
	"phase",
	"power_estimate",
	"power_free_info",
	"power_print",
	"print",
	"print_altname",
	"print_clock",
	"print_delay",
	"print_factor",
	"print_gate",
	"print_io",
	"print_kernel",
	"print_latch",
	"print_level",
	"print_library",
	"print_map_statsprint_state",
	"print_stats",
	"print_value",
	"quit",
	"read_astg",
	"read_blif",
	"read_eqn",
	"read_kiss",
	"read_library",
	"read_pla",
	"read_slif",
	"red_removal",
	"reduce_depth",
	"remove_dep",
	"remove_latches",
	"replace",
	"reset_name",
	"resub",
	"retime",
	"save",
	"set",
	"set_delay",
	"set_state",
	"short_tests",
	"sim_verify",
	"simplify",
	"simulate",
	"source",
	"speed_up",
	"speedup_alg",
	"state_assign",
	"state_minimize",
	"stg_cover",
	"stg_extract",
	"stg_to_astg",
	"stg_to_network",
	"sweep",
	"tech_decomp",
	"time",
	"timeout",
	"unalias",
	"undo",
	"unset",
	"usage",
	"verify",
	"verify_fsm",
	"wd",
	"write_astg",
	"write_bdnet",
	"write_blif",
	"write_eqn",
	"write_kiss",
	"write_pds",
	"write_pla",
	"write_slif",
	"xl_absorb",
	"xl_ao",
	"xl_coll_ck",
	"xl_cover",
	"xl_decomp_two",
	"xl_imp",
	"xl_k_decomp",
	"xl_merge",
	"xl_part_coll",
	"xl_partition",
	"xl_rl",
	"xl_split",
    	NULL
};

char **command_name_completion(const char *text, int start, int end)
{
    return rl_completion_matches(text, command_name_generator);
}

char *command_name_generator(const char *text, int state)
{
    static int list_index, len;
    char *name;

    if (!state) {
        list_index = 0;
        len = strlen(text);
    }

    while ((name = commands[list_index++])) {
        if (strncmp(name, text, len) == 0) {
            return strdup(name);
        }
    }

    return NULL;
}

static int source_sisrc(network)
	network_t **network; {
	char *cmdline;
	char *lib_name;
	char *homefile;
	struct stat home, cur;
	int s1, s2; /* Flags for checking the stat() call */
	int status0, status1, status2, status3, status4, status5;

	lib_name = sis_library();
	cmdline = ALLOC(char, strlen(lib_name) + 20);

	(void) sprintf(cmdline, "source -s %s/.misrc", lib_name);
	status0 = com_execute(network, cmdline);
	(void) sprintf(cmdline, "source -s %s/.sisrc", lib_name);
	status1 = com_execute(network, cmdline);
	status3 = com_execute(network, "source -s ~/.sisrc");

	homefile = util_tilde_expand("~/.misrc");
	s1 = stat(homefile, &home);
	s2 = stat(".misrc", &cur);
	status2 = status4 = TRUE;
	if ((s1 == 0) && (s2 == 0) && (home.st_ino == cur.st_ino)) {
		/* ~/.misrc == .misrc : Source the file only once */
		status2 = com_execute(network, "source -s ~/.misrc");
	} else {
		if (s1 == 0) { /* ~/.misrc exists and can be opened */
			status2 = com_execute(network, "source -s ~/.misrc");
		}
		if (s2 == 0) { /* ./.misrc exists and can be opened */
			status4 = com_execute(network, "source -s .misrc");
		}
	}
	FREE(homefile);

	homefile = util_tilde_expand("~/.sisrc");
	s1 = stat(homefile, &home);
	s2 = stat(".sisrc", &cur);
	status3 = status5 = TRUE;
	if ((s1 == 0) && (s2 == 0) && (home.st_ino == cur.st_ino)) {
		/* ~/.sisrc == .sisrc : Source the file only once */
		status3 = com_execute(network, "source -s ~/.sisrc");
	} else {
		if (s1 == 0) {
			status3 = com_execute(network, "source -s ~/.sisrc");
		}
		if (s2 == 0) {
			status5 = com_execute(network, "source -s .sisrc");
		}
	}
	FREE(homefile);

	FREE(cmdline);
	FREE(lib_name);
	return status0 && status1 && status2 && status3 && status4 && status5;
}

static void usage(prog)
	char *prog; {
	char *lib_name;

	(void) fprintf(miserr, "%s\n", sis_version());
	(void) fprintf(miserr,
			"usage: %s [-sx] [-c cmd] [-f script] [-o file] [-t type] [-T type] [file]\n",
			prog);
#ifdef SIS
	(void) fprintf(miserr, "    -c cmd\texecute SIS commands `cmd'\n");
	(void) fprintf(miserr, "    -f file\texecute SIS commands from a file\n");
#else
	(void) fprintf(miserr,
			"    -c cmd\texecute MIS commands `cmd'\n");
	(void) fprintf(miserr,
			"    -f file\texecute MIS commands from a file\n");
#endif
	(void) fprintf(miserr,
			"    -o file\tspecify output filename (default is -)\n");
	lib_name = sis_library();
#ifdef SIS
	(void) fprintf(miserr, "    -s\t\tsuppress initial 'source %s/.sisrc'\n",
			sis_library());
#else
	(void) fprintf(miserr,
			"    -s\t\tsuppress initial 'source %s/.misrc'\n", sis_library());
#endif
	FREE(lib_name);
#ifdef SIS
	(void) fprintf(miserr,
			"    -t type\tspecify input type (blif, eqn, kiss, oct, pla, slif, or none)\n");
	(void) fprintf(miserr,
			"    -T type\tspecify output type (blif, eqn, kiss, oct, pla, slif, or none)\n");
#else
	(void) fprintf(miserr,
			"    -t type\tspecify input type (bdnet, blif, eqn, oct, pla, or none)\n");
	(void) fprintf(miserr,
			"    -T type\tspecify output type (bdnet, blif, eqn, oct, pla, or none)\n");
#endif
	(void) fprintf(miserr, "    -x\t\tequivalent to '-t none -T none'\n");
	exit(2);
}

static int check_type(s)
	char *s; {
	if (strcmp(s, "bdnet") == 0) {
		return 1;
	} else if (strcmp(s, "blif") == 0) {
		return 1;
	} else if (strcmp(s, "eqn") == 0) {
		return 1;
#ifdef SIS
	} else if (strcmp(s, "kiss") == 0) {
		return 1;
#endif
	} else if (strcmp(s, "oct") == 0) {
		return 1;
	} else if (strcmp(s, "pla") == 0) {
		return 1;
#ifdef SIS
	} else if (strcmp(s, "slif") == 0) {
		return 1;
#endif
	} else if (strcmp(s, "none") == 0) {
		return 1;
	} else {
		(void) fprintf(miserr, "unknown type %s\n", s);
		return 0;
	}
}

int main(argc, argv)
	int argc;char **argv; {
	int c, status, batch, initial_source, initial_read, final_write;
	int quit_flag;
	char readcmd[20], writecmd[20];
	char *dummy, *cmdline, *cmdline1, *infile, *outfile;

	program_name = argv[0];

	quit_flag = -1; /* Quick quit */
	util_getopt_reset();

	if (main_has_restarted) {
		(void) fprintf(stderr, "Restarting frozen image ...\n");
	} else {
		main_has_restarted = 1;
		init_sis();
		network = network_alloc();
	}

	cmdline = util_strsav("");
	(void) strcpy(readcmd, "read_blif");
	(void) strcpy(writecmd, "write_blif");
	infile = "-";
	outfile = "-";
	command_hist = array_alloc(char *, 0);
	initial_source = 1;
	initial_read = 1;
	final_write = 1;
	batch = 0;
	util_getopt_reset();
	while ((c = util_getopt(argc, argv, "c:f:o:st:T:xX:")) != EOF) {
		switch (c) {
		case 'c':
			FREE(cmdline);
			cmdline = util_strsav(util_optarg);
			batch = 1;
			break;

		case 'f':
			FREE(cmdline);
			cmdline = ALLOC(char, strlen(util_optarg) + 20);
			(void) sprintf(cmdline, "source %s", util_optarg);
			batch = 1;
			break;

		case 'o':
			outfile = util_optarg;
			break;

		case 's':
			initial_source = 0;
			break;

		case 't':
			if (check_type(util_optarg)) {
				if (strcmp(util_optarg, "none") == 0) {
					initial_read = 0;
				} else {
					(void) sprintf(readcmd, "read_%s", util_optarg);
				}
			} else {
				usage(argv[0]);
			}
			batch = 1;
			break;

		case 'T':
			if (check_type(util_optarg)) {
				if (strcmp(util_optarg, "none") == 0) {
					final_write = 0;
				} else {
					(void) sprintf(writecmd, "write_%s", util_optarg);
				}
			} else {
				usage(argv[0]);
			}
			batch = 1;
			break;

		case 'x':
			final_write = 0;
			initial_read = 0;
			batch = 1;
			break;

		case 'X':
			/* Handled in previous option loop. */
			break;

		default:
			usage(argv[0]);
		}
	}

	if (!batch) {
		/* interactive use ... */
		if (argc - util_optind != 0) {
			(void) fprintf(miserr, "warning -- trailing arguments ignored\n");
		}

		(void) fprintf(misout, "%s\n", sis_version());
		if (initial_source) {
			(void) source_sisrc(&network);
		}
		rl_attempted_completion_function = command_name_completion;
		do{
			cmdline=readline("sis> ");
			if(cmdline){
				quit_flag=com_execute(&network, cmdline);
				add_history(cmdline);
			}
		}while(quit_flag>=0);
		status = 0;
	} else {

		/* read initial network */
		if (argc - util_optind == 0) {
			infile = "-";
		} else if (argc - util_optind == 1) {
			infile = argv[util_optind];
		} else {
			usage(argv[0]);
		}

		if (initial_source) {
			(void) source_sisrc(&network);
		}

		status = 0;
		if (initial_read) {
			cmdline1 = ALLOC(char, strlen(infile) + 20);
			(void) sprintf(cmdline1, "%s %s", readcmd, infile);
			status = com_execute(&network, cmdline1);
			FREE(cmdline1);
		}

		if (status == 0) {
			status = com_execute(&network, cmdline);
			if ((status == 0 || status == -1) && final_write) {
				cmdline1 = ALLOC(char, strlen(outfile) + 20);
				(void) sprintf(cmdline1, "%s %s", writecmd, outfile);
				status = com_execute(&network, cmdline1);
				FREE(cmdline1);
			}
		}

	}

	FREE(cmdline);
	for (c = array_n(command_hist); c-- > 0;) {
		dummy = array_fetch(char *, command_hist, c);
		FREE(dummy);
	}
	array_free(command_hist);
	/* Value of "quit_flag" is determined by the "quit" command */
	if (quit_flag == -1 || quit_flag == -2) {
		status = 0;
	}
	if (quit_flag == -2) {
		network_free(network);
		end_sis();
	}
	exit(status);
}
